### PR TITLE
Comprehensive Testing for Challenge Collection in Firestore

### PR DESCRIPTION
**Pull request by samymessouak**
This pull request involves implementing and verifying comprehensive test cases to ensure the robustness and correctness of the new challenge collection in Firestore.

## Summary of Changes:

* Added comprehensive test coverage to ensure the functionality and reliability of challenge-related features.
* Validated integration with Firestore and ensured data consistency across operations.

### Key Tests Implemented:

- **ChallengeRepositoryFirestore Tests**
  * Developed tests to validate the integration with the challenge collection.
  * Ensured that challenge creation and deletion functionalities operate seamlessly.

- **ChallengeViewModel Tests**
  * Implemented tests for view-model logic.
  * Verified proper handling of challenge lifecycle operations, including creating and deleting challenges.
  * Confirmed real-time updates with Firestore.

- **UserViewModel Tests**
  * Added test cases to confirm functions like `getOngoingChallenge` and `removeOngoingChallenge` work as expected.
  * Ensured smooth management of ongoing challenges from the user's perspective.

- **UserRepositoryFirestore Tests**
  * Enhanced testing for user-related operations involving challenges.
  * Confirmed correct association of challenges with users and maintained data integrity across updates.

## Impact:

These tests collectively ensure that the new challenge collection is:

* Fully functional
* Reliable
* Integrated seamlessly into the existing system

The test suite provides a solid foundation for future feature expansions and scalability.

image
image
image

### CODE REVIEW

File: app/src/test/java/com/github/se/signify/model/user/UserViewModelTest.kt [hunk 1]
```diff
@@ -87,4 +88,16 @@ class UserViewModelTest {
     userViewModel.removeFriend(currentUserId, friendUserId)
     verify(userRepository).removeFriend(eq(currentUserId), eq(friendUserId), any(), any())
   }
+
```
Comment (SaFouNaldo) [inline app/src/test/java/com/github/se/signify/model/user/UserViewModelTest.kt:91]: The two added tests check if remove & add a challenge is being called in the user repo bu using a mocked model.
Simple and efficient good job

File: app/src/test/java/com/github/se/signify/model/user/UserRepositoryFireStoreTest.kt [hunk 2]
```diff
@@ -607,4 +616,150 @@ class ToDosRepositoryFireStoreTest {
     // Verify that the friendUser update was attempted and failed
     verify(mockFriendUserDocRef).update(eq("friends"), ArgumentMatchers.any(FieldValue::class.java))
   }
+
```
Comment (SaFouNaldo) [inline app/src/test/java/com/github/se/signify/model/user/UserRepositoryFireStoreTest.kt:619]: Test to check if the userRepo acctually changed when adding and removing.
Also tested the cases where it is supposed to fail.
Again nothing complicated simple and efficient. 
Good job

File: app/src/test/java/com/github/se/signify/model/user/UserRepositoryFireStoreTest.kt [hunk 3]
```diff
@@ -60,6 +65,10 @@ class ToDosRepositoryFireStoreTest {
 
     `when`(mockCollectionReference.document(currentUserId)).thenReturn(mockCurrentUserDocRef)
     `when`(mockCollectionReference.document(friendUserId)).thenReturn(mockFriendUserDocRef)
+    `when`(mockFireStore.collection("challenges").document(challengeId1))
```
Comment (SaFouNaldo) [inline app/src/test/java/com/github/se/signify/model/user/UserRepositoryFireStoreTest.kt:68]: ensure that there are no dependecies when testing

File: app/src/test/java/com/github/se/signify/model/user/SaveUserToFireStoreTest.kt [hunk 4]
```diff
@@ -28,7 +28,7 @@ import org.robolectric.RobolectricTestRunner
 import org.robolectric.Shadows.shadowOf
 
 @RunWith(RobolectricTestRunner::class)
-class UserRepositoryFireStoreTest {
+class SaveUserToFireStoreTest {
```
Comment (SaFouNaldo) [inline app/src/test/java/com/github/se/signify/model/user/SaveUserToFireStoreTest.kt:31]: excha,ged the name of the classes for better modularity and readability

File: app/src/test/java/com/github/se/signify/model/challenge/ChallengeViewModelTest.kt [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package com.github.se.signify.model.challenge
```
Comment (SaFouNaldo) [inline app/src/test/java/com/github/se/signify/model/challenge/ChallengeViewModelTest.kt:1]: Test the challenge view model a pretty forward class test

**Review by SaFouNaldo — APPROVED** [review]
Good test PR simple and efficient.
Next step would be to push the logic and the tests in the same branch as stayted by the coaches.
I ofc approve this PR

**Review by ziyadbagh — COMMENTED** [review]
The tests are well-implemented and thoroughly cover the essential Firestore operations for managing challenges—great job on both the new and existing test classes. One note for future improvement: consider adding edge case tests, such as scenarios where both player IDs are empty or null. I’ll go ahead and approve this.

**Review by ziyadbagh — APPROVED** [review]
